src/types.md

```diff
@@ -0,0 +1,183 @@
+# Questions about your types
+
+## My 'class' needs references to lots of other things in order to do its job, and that conflicts with other things which also hold similar references. What do I do?
+
+It's common in C++ to have objects which need to mutate other objects to get
+their work done. Those other objects may be accessed by several different
+objects.
+
+```mermaid
+flowchart LR
+    subgraph Shared functionality
+    important[Important Shared Object]
+    end
+    subgraph ObjectA
+    methodA[Method]
+    refa[Mutable Reference]-->important
+    methodA-. Acts on shared object.->important
+    end
+    subgraph ObjectB
+    refb[Mutable Reference]-->important
+    methodB[Method]
+    methodB-. Acts on shared object.->important
+    end
+    main --> ObjectA
+    main --> ObjectB
+    main-. Calls .-> methodA
+    main-. Calls .-> methodB
+```
+
+You can't have multiple mutable references to a shared object, so what do you do?
+
+First of all, consider moving behavior out of your types. (See
+[the answer about the observer pattern](./codebase.md#the-c-observer-pattern-is-hard-in-rust-what-to-do) and especially
+[the second option described there](./codebase.md#option-2-drive-the-objects-from-the-code-not-the-other-way-round).)
+
+Even in Rust, though, it's still often the best choice to have complex
+behavior as part of types within `impl` blocks. You can still do that -
+but don't _store_ references. Instead, pass them into each function call.
+
+```mermaid
+flowchart LR
+    subgraph Shared functionality
+    important[Important Shared Object]
+    end
+    subgraph ObjectA
+    methodA[Method]
+    methodA-. Acts on shared object.->important
+    end
+    subgraph ObjectB
+    methodB[Method]
+    methodB-. Acts on shared object.->important
+    end
+    main --> ObjectA
+    main --> ObjectB
+    main --> important
+    main-. Passes reference to shared object.-> methodA
+    main-. Passes reference to shared object.-> methodB
+```
+
+Instead of:
+
+```rust
+struct ImportantSharedObject;
+
+struct ObjectA<'a> {
+    important_shared_object: &'a mut ImportantSharedObject,
+}
+
+impl<'a> ObjectA<'a> {
+    fn new(important_shared_object: &'a mut ImportantSharedObject) -> Self {
+        Self {
+            important_shared_object
+        }
+    }
+    fn do_something(&mut self) {
+        // act on self.important_shared_object
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut a = ObjectA::new(&mut shared_thingy);
+    a.do_something();
+}
+```
```

Comment:
I wonder if you can leave out the definition of `ImportantSharedObject` and `ObjectA` and instead just show the `main()` function? I think that makes things clear enough -- the implied interface of the objects is pretty obvious.